include/srf/segment/builder.hpp

```diff
@@ -161,6 +165,53 @@ class Builder final
         return construct_object<NodeTypeT<SinkTypeT, SourceTypeT>>(name, std::forward<ArgsT>(ops)...);
     }
 
+    /**
+     * Instantiate a segment module of `ModuleTypeT`, intialize it, and return it to the caller
+     * @tparam ModuleTypeT Type of module to create
+     * @param module_name Unique name of this instance of the module
+     * @param config Configuration to pass to the module
+     * @return Return a shared pointer to the new module, which is a derived class of SegmentModule
+     */
+    template <typename ModuleTypeT>
+    std::shared_ptr<ModuleTypeT> make_module(std::string module_name, nlohmann::json config = {})
+    {
+        static_assert(std::is_base_of_v<modules::SegmentModule, ModuleTypeT>);
+
+        auto module = std::make_shared<ModuleTypeT>(std::move(module_name), std::move(config));
+        init_module(module);
+
+        return std::move(module);
+    }
+
+    /**
+     * Initialize a SegmentModule that was instantiated outside of the builder.
+     * @param module Module to initialize
+     */
+    void init_module(std::shared_ptr<srf::modules::SegmentModule> module);
+
+    /**
+     * Register an input port on the given module -- note: this in generally only necessary for dynamically
+     * created modules that use an alternate initializer function, which is independent of the derived class.
+     * See: PythonSegmentModule
+     * @param input_name Unique name of the input port
+     * @param object shared pointer to type erased Object associated with 'input_name' on this module instance.
+     */
+    void register_module_input(std::string input_name, std::shared_ptr<segment::ObjectProperties> object);
+
+    /**
+     * Register an outputport on the given module -- note: this in generally only necessary for dynamically
+     * created modules that use an alternate initializer function, which is independent of the derived class.
+     * See: PythonSegmentModule
+     * @param output_name Unique name of the output port
+     * @param object shared pointer to type erased Object associated with 'output_name' on this module instance.
+     */
+    void register_module_output(std::string output_name, std::shared_ptr<segment::ObjectProperties> object);
```

Comment:
I may not fully understand your suggestion here, maybe we should discuss.

In general, module implementations in c++ should use the base class register_input|output interfaces.

These methods on the builder really only exist for dynamic module registration (python). See `pysrf/py_segment_module.hpp`

Reply:
I think this would not be a trivial change.

If we have a derived ModuleBuilder then python initializer functions for modules will have different signatures than standard initializer functions; down casting from a ModuleBuilder to a Builder should work without any issue, but we have some potential for issues with nested module construction.

While we do leave open the possibility of users calling the registration functions from a non-module context, its a case that is caught and throws an appropriate error, and I don't think would be overly confusing.